packages/style-engine/class-wp-style-engine.php

```diff
@@ -261,29 +365,75 @@ public function generate( $block_styles ) {
 	/**
 	 * Default style value parser that returns a CSS ruleset.
 	 * If the input contains an array, it will be treated like a box model
-	 * for styles such as margins and padding
+	 * for styles such as margins and padding.
 	 *
-	 * @param string|array $style_value    A single raw Gutenberg style attributes value for a CSS property.
-	 * @param string       $style_property The CSS property for which we're creating a rule.
+	 * @param string|array $style_value      A single raw Gutenberg style attributes value for a CSS property.
+	 * @param array        $style_definition A single style definition from BLOCK_STYLE_DEFINITIONS_METADATA.
 	 *
 	 * @return array The class name for the added style.
 	 */
-	protected static function get_css_rules( $style_value, $style_property ) {
+	protected static function get_css_rules( $style_value, $style_definition ) {
 		$rules = array();
 
 		if ( ! $style_value ) {
 			return $rules;
 		}
 
+		$style_properties = $style_definition['properties'];
+
 		// We assume box model-like properties.
 		if ( is_array( $style_value ) ) {
 			foreach ( $style_value as $key => $value ) {
-				$rules[ "$style_property-$key" ] = $value;
+				$side_property           = strtr( $style_properties['sides'], array( '$side' => _wp_to_kebab_case( $key ) ) );
```

Comment:
Just wondering if we need this logic to be hard-coded to the placeholder `$side` — would `%s` work and then we could use `sprintf()` for the substitution?

In the case of this `foreach` loop, it looks like it doesn't necessarily matter whether or not the property is a `$side_property` or any other kind of property — just that the `$key` will be substituted into the property name, and the value will be assigned to that key.

While we don't yet have a background block support, I was idly wondering how we might one day implement a background support / background image support, and whether it'd be helpful to have a single `background` support with separate properties for `background-image`, `background-position`, and `background-repeat` — so there's a potential use case for non-side based properties, there. But, totally happy to kick that idea down the road a bit as it's likely a fair ways off!

Reply:
>  it looks like it doesn't necessarily matter whether or not the property is a $side_property or any other kind of property

Fair point. I think I replaced `%s` /`sprintf()` for `$side`, `$slug` etc to make things easier for reading while testing. `%s` and `sprintf()` will work just fine, and, as you say, better suit non-side based properties. 👍

Reply:
Ah, gotcha! I quite like `$slug` if you want to keep the more explicit reading of the sub-string to be substituted (and I see now that's what `WP_Theme_JSON` is already using — apologies, still catching up!)

Reply:
> so there's a potential use case for non-side based properties

I'd argue it's already being used for corners instead of sides for border-radius. Personally, I'd like to see the terminology of "sides" replaced with something that more broadly applies. Perhaps something towards "longhand" given that's sort of what these "sub-properties" are?